### PR TITLE
Check for int64 overflow when converting from float64

### LINKED ISSUE #2165: promql: int overflow not checked
The float64 sample value can overflow an int, even and int64. We blindly convert to int (and not even int64) for the MOD operator in `engine.go`. This silently gives weird results instead of erroring or NaN'ing. It also breaks the test on certain architectures, cf. #2058.

The evalInt method has a similar issue.

evalInt should probably error out (as that's somewhere where an error is expected anyway). MOD should probably yield NaN (similar to MOD by 0).

### PR DESCRIPTION
**Pull request by beorn7**
Fixes #2165 

@TheTincho @brian-brazil @fabxc

### CODE REVIEW

File: promql/engine.go [hunk 1]
```diff
@@ -30,6 +30,18 @@ import (
 	"github.com/prometheus/prometheus/util/stats"
 )
 
+const (
+	// The largest SampleValue that can be converted to an int64 without overflow.
+	maxInt64 model.SampleValue = 9223372036854774784
+	// The smallest SampleValue that can be converted to an int64 without underflow.
+	minInt64 model.SampleValue = -9223372036854775808
+)
```
Comment (fabxc) [inline promql/engine.go]: https://godoc.org/math#pkg-constants

Comment (beorn7) [inline promql/engine.go]: These aren't the constants you are looking for.

Comment (fabxc) [inline promql/engine.go]: My bad, just dived in and didn't check the context properly.

**Comment by fabxc** [comment]
Anything speaking against using https://godoc.org/math#Mod?

**Comment by beorn7** [comment]
My naive expectation was that the `%` operator in PromQL would behave exactly like https://godoc.org/math#Mod . Only when I debugged those recent issues, I realized that the actual implementation first converts everything to integers. I guess there was a rationale behind it. @juliusv do you have historic background on the decision?

Not knowing any other context, I'd prefer the https://godoc.org/math#Mod behavior, but changing it now would arguably be a breaking change as people might (deliberately or accidentally) rely on the rounding-down-to-int behavior.

**Comment by brian-brazil** [comment]
I would have expected this not to round to an integer result, and would consider it a bug that it doesn't. It's low impact so we might want to leave fixing it to 2.0 to be on the safe side.

**Comment by fabxc** [comment]
Every bug is someone's feature. If we all agree that this doesn't match our actual intention, we might not want to drag this around as additional baggage into 2.0.

The fact that it's so low impact is probably a good argument not to worry about it.

File: promql/engine.go [hunk 2]
```diff
@@ -30,6 +30,18 @@ import (
...
+
+// convertibleToInt64 returns true if v does not over-/underflow an int64.
+func convertibleToInt64(v model.SampleValue) bool {
+	return v <= maxInt64 && v >= minInt64
```
Comment (juliusv) [inline promql/engine.go]: Could this be solved simplar without constants by just:

```
return model.SampleValue(int64(v)) == v
```

Like we do it in the chunk encoding code: https://github.com/prometheus/prometheus/blob/master/storage/local/chunk/delta_helpers.go#L81-L84

**Comment by juliusv** [comment]
@beorn7 I'm pretty sure I just used `%` back then because I didn't know about `math.Mod()` and was unaware of the concept of a floating-point mod.

**Comment by beorn7** [comment]
OK, under the given circumstances, it's sane to assume that the `math.Mod` behavior was always the intended one. I doubt that anybody ever relied on the currently implemented behavior.

So yeah, let's go for `math.Mod` and declare the current behavior buggy. I still leave the fix for the int conversion in place, as that still makes sense. But I'll update all the mod calculations to `math.Mod`.

Thanks everybody for your input.

**Comment by beorn7** [comment]
OK, updated. PHAL.

**Comment by juliusv** [comment]
:+1:

**Pull request merged**